### PR TITLE
Add pwalkdir, speed up chcon

### PR DESCRIPTION
**Pull request by kolyshkin**
TL;DR: speed up chcon for up to 1.5x to 2x with Go 1.16+.

1. pkg/pwalkdir: add
    
    This is a copy-paste implementation of pwalk but utilizing
    filepath.WalkDir instead (added in Go 1.16). As WalkDir do not call
    stat(2) on every entry, it is much (2x on my machine) faster than Walk
    (see below).
    
    The idea is to have pwalk and pwalkdir coexist for some time, then
    gradually retire pwalk once Go < 1.16 is unsupported.

2. Chcon: speedup for Go 1.16+
    
    Using the newly added pwalkdir, the benchmark added by and described
    in commit ba956cafe09661021bf shows a considerable speed improvement
    (0.8s with pwalk, 0.5s with pwalkdir, 0.9s with chcon from GNU coreutils).

### CODE REVIEW

**Comment by kolyshkin** [comment]
A few non-scientific benchmarks (testdata is Linux kernel source tree, find is for reference only)

| Op | Time |
| --- | --- |
| find -print >/dev/null | 0.19s |
| chcon -R | 0.92s |
| Before this PR (pwalk) | 0.83s |
| This PR (pwalkdir) + go 1.16 | 0.51s |
| This PR (pwalkdir) + go 1.17rc2 | 0.45s |

**Comment by thaJeztah** [comment]
Nice! Will try to do a review soon, but ping me otherwise 😅

**Review by rhatdan — APPROVED** [review]
LGTM

**Comment by rhatdan** [comment]
Awesome work @kolyshkin

File: pkg/pwalkdir/pwalkdir.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+import (
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"runtime"
+	"sync"
+)
+
+type WalkFunc = fs.WalkDirFunc
```
Comment (thaJeztah) [inline pkg/pwalkdir/pwalkdir.go:13]: Probably want to rename the alias as well. `WalkDirFunc` is not a drop-in replacement for `pwalk.WalkFunc` (different signature), so perhaps better to rename?

```suggestion
type WalkDirFunc = fs.WalkDirFunc
```

Comment (thaJeztah) [reply pkg/pwalkdir/pwalkdir.go:13]: Or should we remove the alias altogether? The signature has `fs.Direntry` in it, so consumers would still have to add `import "io/fs"` (so no extra convenience there)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Spotted a typo, and one question/suggestion (rename or drop the `WalkFunc` type alias); let me know what you think

File: pkg/pwalkdir/pwalkdir.go (hunk 1 above)
Comment (kolyshkin) [reply pkg/pwalkdir/pwalkdir.go:13]: Yep, I think it is best to remove the alias.

Comment (kolyshkin) [reply pkg/pwalkdir/pwalkdir.go]: removed

File: pkg/pwalkdir/pwalkdir.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+// Walk is a wrapper for filepath.WalkDir which can call multiple walFn
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply pkg/pwalkdir/pwalkdir.go]: fixed

**Comment by thaJeztah** [comment]
Looks like you forgot to remove some references that used the alias;

```
  level=error msg="Running error: buildir: analysis skipped: errors in package: [/home/runner/work/selinux/selinux/pkg/pwalkdir/pwalkdir_test.go:124:42: undeclared name: WalkFunc /home/runner/work/selinux/selinux/pkg/pwalkdir/pwalkdir_test.go:127:34: undeclared name: WalkFunc /home/runner/work/selinux/selinux/pkg/pwalkdir/pwalkdir_test.go:141:8: undeclared name: WalkFunc /home/runner/work/selinux/selinux/pkg/pwalkdir/pwalkdir_test.go:153:24: cannot use filepath.WalkDir (value of type func(root string, fn fs.WalkDirFunc) error) as walkerFunc value in struct literal /home/runner/work/selinux/selinux/pkg/pwalkdir/pwalkdir_test.go:154:21: cannot use Walk (value of type func(root string, walkFn fs.WalkDirFunc) error) as walkerFunc value in struct literal]"
```

**Comment by kolyshkin** [comment]
> Looks like you forgot to remove some references that used the alias

🤦🏻 never trust 3am edits! Fixed.

**Comment by thaJeztah** [comment]
> 🤦🏻 never trust 3am edits! Fixed

More than familiar with those 😂

**Review by thaJeztah — APPROVED** [review]
LGTM!

**Comment by thaJeztah** [comment]
@rhatdan ptal
